source/mainnet/smart-contracts/onboarding-guide-ethereum-developers/faq.rst

```diff
@@ -0,0 +1,792 @@
+
+    **Smart contract errors**
+
+    An example of a smart contract error is shown below:
+
+    .. code-block:: console
+
+        Error: Updating contract instance failed:
+        'smash' in 'PiggyBank' at {"index":12345,"subindex":0} failed with code -1.
+
+    .. image:: ../tutorials/piggy-bank/images/pb_tutorial_27.png
+        :width: 100 %
+
+    You can find the meaning of the piggy bank error codes in this :ref:`section<piggy-bank-smash-error>`.
+    This error was thrown by the smart contract. The error codes start from ``-1`` and go downwards.
+    ``-1`` is the first error code and corresponds to ``NotOwner``.
+    In contrast, errors from the ``concordium-std`` crate start from ``i32::MIN`` and go upwards.
+
+.. dropdown:: Is there a smart contract code linter?
+
+    Yes. You can use the ``fmt`` and the ``cargo clippy`` linter tools as described in the `README <https://github.com/Concordium/concordium-rust-smart-contracts>`_.
+
+.. dropdown:: Are tests executed in parallel or sequentially?
+
+    Tests run in parallel by default with the following two commands
+
+    .. code-block:: console
+
+        $cargo concordium test
+
+    .. code-block:: console
+
+        $cargo test
+
+    You can use the below command for consecutive execution of the tests.
+
+    .. code-block:: console
+
+        $cargo test -- --test-threads=1
+
+.. dropdown:: How do I embed a schema into a smart contract? Why do I need a schema? How can I provide the input parameters as a JSON object and get the output parameters in a human-readable format when using the `concordium-client`?
+
+    A :ref:`smart contract schema<contract-schema>` is a description of how to represent
+    bytes in a more structured representation. It is used by
+    external tools (such as the ``concordium-client``) when displaying the return value of a receive
+    function and for specifying the input parameters to a receive
+    function using a structured representation, such as JSON.
+
+    The smart contract function needs to define a ``parameter`` and a `return_value``
+    so that a schema can be generated for the input parameters and the output parameters, respectively.
+
+    .. code-block:: rust
+        :emphasize-lines: 4, 5
+
+        #[receive(
+            contract = "CIS2-wCCD",
+            name = "balanceOf",
+            parameter = "ContractBalanceOfQueryParams",
+            return_value = "ContractBalanceOfQueryResponse"
+        )]
+        fn contract_balance_of<S: HasStateApi>(
+            ctx: &impl HasReceiveContext,
+            host: &impl HasHost<StateImplementation, StateApiType = S>,
+        ) -> ContractResult<ContractBalanceOfQueryResponse> { ... }
+
+
+    The schema can either be embedded into the smart contract or generated
+    as a separate file.
+
+    **Option 1 (Schema embedded into a smart contract)**
+
+    .. code-block:: console
+
+        $cargo concordium build -e
```

Comment:
Perhaps we should use the long option here? `--schema-embed`

Reply:
I added the long version as follows:
 
    .. code-block:: console

        $cargo concordium build --schema-embed

    The shorter version of the above command is as follows:

    .. code-block:: console

        $cargo concordium build -e

Reply:
Okay, I don't think explaining the shorthand is necessary to explain. They can find that with `cargo concordium build --help`. But if you want to keep it, then also describe the shorthand for the schema-out (`-s`) 😊

Reply:
The screenshot below uses  the `cargo concordium build -e` command  that is why I wanted to keep the command initially (but ended up having to update the below screenshot now instead).

It is never a bad idea to explain too much so I added the shorthand  `schema-out (-s)` to the below command as well and updated the screenshot  :).